cpp/src/interop/to_arrow_device.cu

```diff
@@ -594,49 +587,38 @@ struct ArrowDeviceArrayPrivateData {
 void ArrowDeviceArrayRelease(ArrowArray* array)
 {
   auto private_data = reinterpret_cast<ArrowDeviceArrayPrivateData*>(array->private_data);
-  cudaEventDestroy(private_data->sync_event);
+  CUDF_CUDA_TRY(cudaEventDestroy(private_data->sync_event));
```

Comment:
I'm not sure how I feel about the release callback function being able to throw an exception. Though I can't think of a better solution

Reply:
I agree, this seems problematic. The case I'm most concerned about is, what happens if the produced array device array is passed to a C library and then throws? The fact that the release callback has no way to inform the caller that something failed in the release seems like a potentially problematic oversight in the spec, no?

Reply:
You are quite right. 
This issue came up some time ago with `cudaFree()` calls in the `rmm::device_buffer()` destructor. 
In that case, we do not throw -- at least in Release builds. 
In debug mode, if an error occurs it is written to `std::cerr` an `assert()` is called.
Reference: https://github.com/rapidsai/rmm/blob/9e6db746f1a4a6361fb9fadf381f749dc52faaea/include/rmm/detail/error.hpp#L172-L188

I think we should wrap this with `RMM_ASSERT_CUDA_SUCCESS` instead of `CUDF_CUDA_TRY` here.

Reply:
Updated to call `RMM_ASSERT_CUDA_SUCCESS` instead here.
libcudf does **not** have a similar macro as far as I can tell https://github.com/rapidsai/cudf/blob/eaae68d8b099e90a2e3bcc968f98c652d36bb844/cpp/include/cudf/utilities/error.hpp#L263

Reply:
That looks good, thanks!

Reply:
I'd be curious to hear from @paleolimbot or @kkraus14 about whether there was ever discussion when creating the Arrow specification of defining the release callback as returning an (integer) error code. On one hand I see the rationale for not, since you could argue that this is essentially like the object's destructor and there are good reasons that destructors cannot throw exceptions relating to improper stack unwinding etc. On the other hand, due to the nature of private data being an opaque data pointer that could hold anything, the release callback could be doing any number of thing beyond freeing memory; in fact, the device interface's usage of the release callback to destroy the sync event is precisely such an example. It seems like allowing some informational return that the caller may be able to act upon (even if it's just by throwing an exception later) is potentially valuable.